### PR TITLE
Add ability to correct intensity of cuts

### LINKED ISSUE #360: Incorrect number of bins calculated for some workspaces
Some non-PSD workspaces (.e.g. [this one](https://github.com/mantidproject/mslice/files/2249440/MAR18120_Ei39.53meV.nxspe.zip)) are treated as having the wrong number of bins.

The Energy axis is saved in workspace.limits when the workspace is loaded as `-70, 40, 0.20992` (min, max, step size). Later, these values are used to calculate the number of steps:
```
 (40 - - 70) / 0.20092 = 524.009
```
Which is then rounded up to 525. For most workspaces, this is correct behaviour, but in this case the number of bins is 524. This can cause errors e.g. when changing the intensity, as the number of bins in the workspace is different to what is expected.

### LINKED ISSUE #457: Ability to make cuts with different "intensity" types
At the moment, the user can view slices with the colour map set as different types of `Intensity` - e.g. the default (`S(q,w)`) and also `Chi''(q,w)`, `GDOS` etc. However, cuts (which only operate on the underlying workspace) are only of `S(q,w)` (because the underlying workspace is only of S(q,w)). This is particularly glaring for the `Interactive Cuts` feature.

It would thus be good to have the ability to make cuts with the different types of `Intensity`.

### LINKED ISSUE #641: Adding Bragg peaks when interactive cut window still open causes error
**Describe the bug**
When an interactive cut window is open, if you try and add a Bragg peak to the slice plot it will not be added, but will still be marked as there. If you then close the interactive cut and try to remove the Bragg peak you get an error.

**To Reproduce**
1. In the Workspace Manager tab of the MSlice interface select the workspace MAR21335_Ei60meV
2. Navigate to the ``Slice`` tab of the ``Workspace Manager`` tab
3. Click ``Display`` in the ``Slice`` tab without changing the default values
4. On the slice plot, select ``Interactive Cuts``
5. Use the cursor to select a rectangular region in the slice plot. A second window with a cut plot should open.
6. Navigate to the ``Information`` menu on the slice plot
7. Select ``Aluminium`` from the submenu for Bragg peaks
8. The Bragg peaks are not displayed but are marked as selected in the menu
9. Close the window with the cut
10. An error message appears (ValueError)

**Expected behavior**
In step 8. the Bragg peaks should get displayed and after step 9. there should not be an error message.

**Mantid Version (please complete the following information):**
 - Mantid Version 6.2

### PR DESCRIPTION
**Pull request by MialLewis**
**Description of work:**
This change adds the ability to correct the intensity of cuts. This provides the following functionality:
- Intensity corrections made to slice's now follow through to interactive cuts taken from the slice.
- On cut plots the `intensity` menu can now be used to correct intensity as per previous slice functionality.

For consistency, the code pattern used to correct intensity for slice plots has largely been reused for cut plots.

**To test:**
Being wary of existing issues which will be fixed in subsequent PRs (#809, #810), follow steps 1-10 below for the `HYS_63373_4pixel`, `MAR25087_Ei40.00meV` and `MAR21335_Ei60meV` datasets.

**SLICE/ICUT TEST**
1. Load dataset.
2. Calculate projection if pixel workspace, using `DeltaE` as `u1`.
3. Display slice.
4. Correct the intensity of the slice using the `Intensity` drop down. Check each intensity, noting that GDOS is currently disabled and displays an explanatory tooltip. (You can type 100 for use a valid field for the temperature)
5. With the intensity corrected to `chi`, use the `Interactive Cut` feature to plot a cut over a region of interest around 0meV. Ensure the cut is taken along `|Q|`, swap the axis if necessary.
6. Observe that the 'dip' around 0meV apparent on the slice, is also apparent on the `icut`.
8. On the slice, toggle `Interactive Cuts`. This will convert the previously plotted cut from an `icut` to a regular cut plot.
9. On this cut plot, observe that the intensity checked on the `Intensity` drop down menu is the same as that previously selected on the parent slice.
10. Using the `Intensity` drop down menu, cycle through intensity correction options and observe behaviour.
11. Close slice and cut plot

**MULTI CUT TEST**
12. With the `MAR25087_Ei60.00meV` dataset selected, plot a cut along `DeltaE` from -5 to 5, over `|Q|` from 0 to 10, with a width of 2.5.
13. Repeat this for the `MAR25087_Ei40.00meV` dataset, though ensure to `plot over`
14. Correct the intensity to `chi`.
15. Observe that for each dataset loaded, you are prompted to provide a sample temperature.
16. Observe that all cuts on the plot are corrected for intensity.
17. Using the `HYS_63373_4pixel` dataset, plot over a set of cuts along `DeltaE` from -0.5 to 0.5, from 1.7 to 2.
18. Observe that you are prompted to enter a sample temperature for the new dataset.
19. Correct the intensity back to `S(Q,E)` and back to `chi`, inferring that the initial line plotted over was plotted with intensity corrected with the checked correction.
20. Do not close plot, as it is used for the following test.

**SCRIPTING TEST**
21. With intensity corrected to `chi`, using the `file` drop down click `generate script to clipboard`.
22. Close the plot
23. Paste the generated script into the mslice cli and press enter.
24. Observe that the plot has been recreated successfully

Resolves #457.
Fixes #360.
Fixes #641.

### CODE REVIEW

File: mslice/tests/pixel_workspace_test.py [hunk 1]
```diff
@@ -75,8 +75,8 @@ def test_add_list(self):
         result = self.workspace + list_to_add
         result = result.get_signal()
         self.assertAlmostEqual(0, result[0][0], 8)
-        self.assertAlmostEqual(13, result[1][47], 8)
-        self.assertAlmostEqual(35, result[3][52], 8)
+        self.assertAlmostEqual(59, result[1][47], 8)
+        self.assertAlmostEqual(84, result[3][52], 8)
```
Comment (MialLewis) [inline mslice/tests/pixel_workspace_test.py:79]: I've had to change this unit test, due to the changes in `apply_with_corrected_shape` in `numpy_helper.py`. The `apply_with_corrected_shape` function is intended to allow workspaces to be scaled by `numpy` arrays even if the `numpy` array resulting from a workspace cannot be scaled by the desired `numpy` array as per `numpy`'s broadcasting rules. When adding the list to a workspace, as per this unit test, the `apply_with_corrected_shape` function previously would alter the two intial arrays, despite them both being compatible as per numpy's broadcasting rules. This doesn't seem right to me, so I've changed it. As far as I can tell, this change does not effect any existing operations done in the mslice code base.

File: mslice/tests/histogram_workspace_test.py [hunk 2]
```diff
@@ -65,11 +65,11 @@ def test_add_list(self):
         result = self.workspace + list_to_add
         result = result.get_signal()
 
-        line = np.multiply(np.linspace(0, 9, 10), 10)
-        expected_values = np.empty((10, 10))
-        for i in range(10):
-            expected_values[i] = line
-        np.testing.assert_array_almost_equal(expected_values, result, 8)
+        ws_representative_array = np.linspace(0, 99, 100).reshape(10, 10).T
```
Comment (MialLewis) [inline mslice/tests/histogram_workspace_test.py:68]: See related comment on `pixel_workspace_test,py`. I now directly compare the `workspace + list` operation to the equivalent `numpy array + list`. To me, these should be equal - they previously were not.

File: mslice/util/numpy_helper.py [hunk 3]
```diff
@@ -9,18 +10,48 @@ def apply_with_corrected_shape(method, shape_array, array, exception):
...
-    if array.shape[1] == 1:
-        array = np.transpose(array)
-    if shape_array.shape[0] in array.shape and shape_array.shape[0] != 1:  # array matches the wrong (leftmost) axis
-        return apply_with_swapped_axes(method, shape_array, array)
-    elif shape_array.shape[1] == array.shape[1]:
-        return method(shape_array, array)
-    else:
-        raise exception
+    try:
+        return_value = method(shape_array, array)
```
Comment (MialLewis) [inline mslice/util/numpy_helper.py:14]: Makes sense to me to first try the method with the existing arrays, before transforming them. This does change the results of some existing tests, see `test_add_list` in `pixel_workspace_test.py` and `histogram_workspace_test.py`.

**Review by mducle — COMMENTED** [review]
I've tested the functionality and it all looks good and works as described. I haven't had time to read through the code though.

Edit: Ideally we should also re-enable the `GDOS` option in the near-ish future (maybe for Mantid 6.5, due in Oct/Nov? or if not by 6.6 due in Feb 2023) but that should be in another issue/PR.

File: mslice/tests/pixel_workspace_test.py (hunk 1 above)
Comment (SilkeSchomann) [reply mslice/tests/pixel_workspace_test.py:79]: That sounds reasonable. If numpy's broadcasting rules can be applied to the initial arrays they should not get altered.

**Review by SilkeSchomann — CHANGES_REQUESTED** [review]
When testing I've experienced two problems:
1. When plotting Bragg peaks on icuts MSlice crashes. This only happens when there should be a Bragg peak available. As soon as the icut is converted to a regular cut plot Bragg peaks work as normal.
2. Running the generated script according to step 23 takes very long and does not display the plot window.

**Comment by MialLewis** [comment]
> When testing I've experienced two problems:
> 
> 1. When plotting Bragg peaks on icuts MSlice crashes. This only happens when there should be a Bragg peak available. As soon as the icut is converted to a regular cut plot Bragg peaks work as normal.
> 2. Running the generated script according to step 23 takes very long and does not display the plot window.

Do you experience (2) consistently?

**Comment by SilkeSchomann** [comment]
> > When testing I've experienced two problems:
> > 
> > 1. When plotting Bragg peaks on icuts MSlice crashes. This only happens when there should be a Bragg peak available. As soon as the icut is converted to a regular cut plot Bragg peaks work as normal.
> > 2. Running the generated script according to step 23 takes very long and does not display the plot window.
> 
> Do you experience (2) consistently?

More often than not. The plot was displayed in only one attempt after waiting for a long time.

**Comment by MialLewis** [comment]
> > > When testing I've experienced two problems:
> > > 
> > > 1. When plotting Bragg peaks on icuts MSlice crashes. This only happens when there should be a Bragg peak available. As soon as the icut is converted to a regular cut plot Bragg peaks work as normal.
> > > 2. Running the generated script according to step 23 takes very long and does not display the plot window.
> > 
> > 
> > Do you experience (2) consistently?
> 
> More often than not. The plot was displayed in only one attempt after waiting for a long time.

I can't replicate this, either in `debug` or `release` mode. Are the data files stored locally?

Could you run the below script and let me know how it behaves:
`# Python Script Generated by Mslice on 2022-08-22 10:47:16
import mslice.cli as mc
import mslice.plotting.pyplot as plt

ws_MAR21335_Ei60meV = mc.Load(Filename='C:/Users/wym92549/Downloads/mslice data\MAR21335_Ei60meV.nxs', OutputWorkspace='MAR21335_Ei60meV')
ws_MAR25087_Ei40_00meV = mc.Load(Filename='C:/Users/wym92549/Downloads/mslice data\MAR25087_Ei40.00meV.nxspe', OutputWorkspace='MAR25087_Ei40.00meV')
HYS_63373_4pixel = mc.Load(Filename='C:/Users/wym92549/Downloads/mslice data\HYS_63373_4pixel.nxspe', OutputWorkspace='HYS_63373_4pixel')
ws_HYS_63373_4pixel_EQ = mc.MakeProjection(InputWorkspace=HYS_63373_4pixel, Axis1='DeltaE', Axis2='|Q|')

fig = plt.gcf()
fig.clf()
ax = fig.add_subplot(111, projection="mslice")
cut_ws_0 = mc.Cut(ws_MAR21335_Ei60meV, CutAxis="DeltaE,-5.0,5.0,0.29411765933036804", IntegrationAxis="|Q|,0.0,2.5,0.0", NormToOne=False, IntensityCorrection="dynamical_susceptibility", SampleTemperature=100.0)
ax.errorbar(cut_ws_0, label="MAR21335_Ei60meV 0.00<|Q|<2.50", color="#1f77b4", marker="o", ls="-", lw=1.5)

cut_ws_1 = mc.Cut(ws_MAR21335_Ei60meV, CutAxis="DeltaE,-5.0,5.0,0.29411765933036804", IntegrationAxis="|Q|,2.5,5.0,0.0", NormToOne=False, IntensityCorrection="dynamical_susceptibility", SampleTemperature=100.0)
ax.errorbar(cut_ws_1, label="MAR21335_Ei60meV 2.50<|Q|<5.00", color="#ff7f0e", marker="o", ls="-", lw=1.5)

cut_ws_2 = mc.Cut(ws_MAR21335_Ei60meV, CutAxis="DeltaE,-5.0,5.0,0.29411765933036804", IntegrationAxis="|Q|,5.0,7.5,0.0", NormToOne=False, IntensityCorrection="dynamical_susceptibility", SampleTemperature=100.0)
ax.errorbar(cut_ws_2, label="MAR21335_Ei60meV 5.00<|Q|<7.50", color="#2ca02c", marker="o", ls="-", lw=1.5)

cut_ws_3 = mc.Cut(ws_MAR21335_Ei60meV, CutAxis="DeltaE,-5.0,5.0,0.29411765933036804", IntegrationAxis="|Q|,7.5,10.0,0.0", NormToOne=False, IntensityCorrection="dynamical_susceptibility", SampleTemperature=100.0)
ax.errorbar(cut_ws_3, label="MAR21335_Ei60meV 7.50<|Q|<10.00", color="#d62728", marker="o", ls="-", lw=1.5)

cut_ws_4 = mc.Cut(ws_MAR25087_Ei40_00meV, CutAxis="DeltaE,-5.0,5.0,0.4000000059604645", IntegrationAxis="|Q|,0.0,2.5,0.0", NormToOne=False, IntensityCorrection="dynamical_susceptibility", SampleTemperature=120.0)
ax.errorbar(cut_ws_4, label="MAR25087_Ei40.00meV 0.00<|Q|<2.50", color="#9467bd", marker="o", ls="-", lw=1.5)

cut_ws_5 = mc.Cut(ws_MAR25087_Ei40_00meV, CutAxis="DeltaE,-5.0,5.0,0.4000000059604645", IntegrationAxis="|Q|,2.5,5.0,0.0", NormToOne=False, IntensityCorrection="dynamical_susceptibility", SampleTemperature=120.0)
ax.errorbar(cut_ws_5, label="MAR25087_Ei40.00meV 2.50<|Q|<5.00", color="#8c564b", marker="o", ls="-", lw=1.5)

cut_ws_6 = mc.Cut(ws_MAR25087_Ei40_00meV, CutAxis="DeltaE,-5.0,5.0,0.4000000059604645", IntegrationAxis="|Q|,5.0,7.5,0.0", NormToOne=False, IntensityCorrection="dynamical_susceptibility", SampleTemperature=120.0)
ax.errorbar(cut_ws_6, label="MAR25087_Ei40.00meV 5.00<|Q|<7.50", color="#e377c2", marker="o", ls="-", lw=1.5)

cut_ws_7 = mc.Cut(ws_MAR25087_Ei40_00meV, CutAxis="DeltaE,-5.0,5.0,0.4000000059604645", IntegrationAxis="|Q|,7.5,10.0,0.0", NormToOne=False, IntensityCorrection="dynamical_susceptibility", SampleTemperature=120.0)
ax.errorbar(cut_ws_7, label="MAR25087_Ei40.00meV 7.50<|Q|<10.00", color="#7f7f7f", marker="o", ls="-", lw=1.5)

cut_ws_8 = mc.Cut(ws_HYS_63373_4pixel_EQ, CutAxis="DeltaE,-0.5,0.5,0.05000000074505806", IntegrationAxis="|Q|,1.7,2.0,0.0", NormToOne=False, IntensityCorrection="dynamical_susceptibility", SampleTemperature=110.0)
ax.errorbar(cut_ws_8, label="HYS_63373_4pixel_EQ 1.70<|Q|<2.00", color="#bcbd22", marker="o", ls="-", lw=1.5)

from mslice.cli.helperfunctions import hide_a_line_and_errorbars, append_visible_handle_and_label
from mslice.util.compat import legend_set_draggable

# hide lines, errorbars, and legends
handles, labels = ax.get_legend_handles_labels()
visible_handles = []
visible_labels = []

append_visible_handle_and_label(visible_handles, handles, visible_labels, labels, 0)

append_visible_handle_and_label(visible_handles, handles, visible_labels, labels, 1)

append_visible_handle_and_label(visible_handles, handles, visible_labels, labels, 2)

append_visible_handle_and_label(visible_handles, handles, visible_labels, labels, 3)

append_visible_handle_and_label(visible_handles, handles, visible_labels, labels, 4)

append_visible_handle_and_label(visible_handles, handles, visible_labels, labels, 5)

append_visible_handle_and_label(visible_handles, handles, visible_labels, labels, 6)

append_visible_handle_and_label(visible_handles, handles, visible_labels, labels, 7)

append_visible_handle_and_label(visible_handles, handles, visible_labels, labels, 8)

legend_set_draggable(ax.legend(visible_handles, visible_labels, fontsize='medium'), True)

ax.set_title('')
ax.set_ylabel(r'Signal/#Events')
ax.set_xlabel(r'Energy Transfer (meV)')
ax.set_ylim(bottom=-0.5598348000024196, top=11.756530800050811)
ax.set_xlim(left=-5.338235294117648, right=5.338235294117648)
mc.Show()
`

**Comment by MialLewis** [comment]
> When testing I've experienced two problems:
> 
> 1. When plotting Bragg peaks on icuts MSlice crashes. This only happens when there should be a Bragg peak available. As soon as the icut is converted to a regular cut plot Bragg peaks work as normal.
> 2. Running the generated script according to step 23 takes very long and does not display the plot window.

I also can't replicate `1.`, which isn't a surprise as I checked it during developed. We'll have to convene tomorrow and try to find out what's causing these differences!

**Comment by SilkeSchomann** [comment]
I've tried out the last script and it runs fine. No delays or missing plot window.

**Comment by SilkeSchomann** [comment]
> 

I can still reproduce the crash caused by overplotting an icut with Bragg peaks using MAR21335_Ei60meV.nxs.

**Comment by MialLewis** [comment]
> > 
> 
> I can still reproduce the crash caused by overplotting an icut with Bragg peaks using MAR21335_Ei60meV.nxs.

Could you provide the stack trace, I still can't reproduce.

**Comment by SilkeSchomann** [comment]
> > > 
> > 
> > 
> > I can still reproduce the crash caused by overplotting an icut with Bragg peaks using MAR21335_Ei60meV.nxs.
> 
> Could you provide the stack trace, I still can't reproduce?

```
Traceback (most recent call last):
  File "C:\Users\eoc57742\mslice\mslice\plotting\plot_window\overplot_interface.py", line 39, in toggle_overplot_line
    plotter_presenter.add_overplot_line(plot_handler.ws_name, key, recoil, cif_file, plot_handler.e_log,
  File "C:\Users\eoc57742\mslice\mslice\plotting\plot_window\cut_plot.py", line 707, in e_log
    if self._e_axis_is_y():
  File "C:\Users\eoc57742\mslice\mslice\plotting\plot_window\cut_plot.py", line 713, in _e_axis_is_y
    ws = get_workspace_handle(self.ws_name)
  File "C:\Users\eoc57742\mslice\mslice\models\workspacemanager\workspace_provider.py", line 16, in get_workspace_handle    raise KeyError('workspace %s could not be found.' % workspace_name)
KeyError: 'workspace MAR21335_Ei60meV_cut(-10.194,7.584) could not be found.'
```

**Comment by MialLewis** [comment]
> plot_handler.e_lo

The functions related to `e_log` don't exist in the latest commit. I removed them at some point (in a poorly named commit) as I realised Bragg peaks will only ever apply when the `e_axis` is the `y_axis`, making the functions redundant. Could you check which commit you have checked out?

**Comment by SilkeSchomann** [comment]
> > plot_handler.e_log
> 
> The functions related to `e_log` don't exist in the latest commit. I removed them at some point (in a poorly named commit) as I realised Bragg peaks will only ever apply when the `e_axis` is the `y_axis`, making the functions redundant. Could you check which commit you have checked out?

I had used test-pr to merge. I've just done the same on a different PC and cannot see the problem anymore.

**Review by SilkeSchomann — APPROVED** [review]
Works fine, good job!

**Pull request merged**